### PR TITLE
Error out when unsupported kwargs are passed to Scale.

### LINKED ISSUE #5619: Symlog linear region
I recently posted [this](http://stackoverflow.com/questions/33935758/matplotlib-symlog-linear-region) question on SO, about what I believe is an incorrect behavior of setting `{x/y}scale('symlog')` in combination with a `linthres{x/y}`. For example:

```python
import numpy as np
import matplotlib.pylab as pl

z = np.linspace(0,1,100)
x = np.ones_like(z)
x[0] = 0

pl.figure()
pl.plot(x, z, '-')
pl.yscale('symlog', linthresy=1e-2)
pl.xlim(-0.1,1.1)
```

Produces the image shown below. I would expect a linear region from 0 to 1e-2, and a log region containing 1e-2, 1e-1, 1. Why isn't this happening? Is this a bug?

figure_1

### LINKED ISSUE #14878: plt.yscale doesn't throw warning with invalid kwarg
Similar to #5619 where when using symlog where `plt.yscale('symlog', linthresy=0.01)` , the threshold ends up as the default value. 

Seems that should've been `plt.yscale('symlog', linthreshy=0.01)` but an error wasn't raised when I put in an invalid kwarg. Minimum working example shown below

```python
import numpy as np
import matplotlib.pyplot as plt

x = np.linspace(-1, 1, 1000)

y = 10 * x

plt.subplot(3, 1, 1)
plt.plot(x, y, 'x')
plt.title('Standard symlog')
plt.yscale('symlog')


plt.subplot(3, 1, 2)
plt.plot(x, y, 'x')
plt.title('yscale with linthresy as a kwarg')
# Should throw warning as i gave it a invalid kwarg
plt.yscale('symlog', linthresy=0.01)

plt.subplot(3, 1, 3)
plt.plot(x, y, 'x')
plt.title('yscale with linthreshy as a kwarg')
plt.yscale('symlog', linthreshy=0.01)

plt.show()
```

image

### PR DESCRIPTION
**Pull request by anntzer**
No deprecation period, both because 6357245 likewise added the exception
for LogScale without deprecation, and because this is by far most likely
to be catching bugs rather than real uses.

Closes #5619, closes  #14878.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by jklymak** [comment]
Is this something we've decided?  Because there are lots of places where we don't check for valid kwargs.

**Comment by anntzer** [comment]
If you look at the linked issues it's pretty clearly a good way to shoot oneself in the foot.
If you're worried about implementation complexity I have something nicer coming up but preferably as a separate PR (well, depends on how much stuff you want to review at once); basically I think the scale classes should just always have the same kwargs (`base, nonpos, ...`) regardless of the axis rather than having `basex, nonposx, ..., basey, nonposy`) and then we can just have normal constructors with normal signatures (`def __init__(self, *, base=10, nonpos='mask', ...):`) and not do any of this ridiculous dance.
(The only reason there's basex/basey is because you can pass both of them simultaneously to `loglog` (`loglog(..., basex=3, basey=5)`) but loglog already needs to figure out what kwargs it passes to what Scale class, so it may as well also normalize the kwargs to the suffixless versions.)
The main complexity causing this to go to a separate PR is just handling the deprecation there...

**Comment by ImportanceOfBeingErnest** [comment]
Breaking user code is not so much an issue as breaking downstream packages. No developper is safe from using the API in an unintended way and writing seemingly sensical code. So as usual with API changes, a deprecation seems useful (similar to how matplotlib would appreciate upstream packages doing the same).

**Comment by anntzer** [comment]
I... don't understand what "sufficiently correct for the user" means.  If you look at the linked issues, I am quite confident that there must be users who are currently passing linthresy instead of linthreshy but actually *not* realizing that they made a mistake, and are therefore generating wrong graphs (note that the older of the two issues has been opened for years and probably none of the reviewers who had a quick look at it noticed the typo from a quick review of the code).  Which may be sufficiently correct for them in the sense that they can use them and publish them and put them in presentations and whatnot, but are still wrong.
In that sense I consider this PR to be more of a bugfix than an API change.

**Comment by ImportanceOfBeingErnest** [comment]
Let me give an example for "sufficiently correct for the user". Previously `plt.show(*args, **kwargs)` took positional arguments. This let a lot of people to write code like `plt.show(fig)`. This does not make any sense, but returns the correct and expected output in all cases where there is only one figure in pyplot anyways. There are a lot of packages out there which still have such code in them; but matplotlib did the correct thing, deprecating the use of positional arguments, such that package maintainers have the chance to correct their mistake without any code actually breaking.

**Comment by anntzer** [comment]
As you mentioned, plt.show(fig) did the correct thing even though it should not be called that way.  My claim is that when an incorrect keyword argument is passed to SymLogScale, it's most likely *not* doing the correct thing, as shown by the bug reports.

**Comment by ImportanceOfBeingErnest** [comment]
This applies to any keyword argument:

    ax.set_yscale('linear', produce_a_nice_plot=True)

**Comment by anntzer** [comment]
Obviously it is possble to use a footgun and shoot in the air rather than in one's foot.  Again I would guess more people are shooting in their foot with it rather than in the air.

Let's see if any other devs want to chime in with their opinion?

**Comment by ImportanceOfBeingErnest** [comment]
Maybe I missed it, so what is the drawback of doing the deprecation with a warning as usual? (A warning by itself should already be enough to keep one's feet healthy)

**Comment by jklymak** [comment]
I guess I think that the general idea that folks should not be allowed to inadvertently make incorrect plots is a fine line to make when deciding whether to check kwargs...

But,... 1) I think this scale is terrible, and I'm not convinced 90% of the people who use it (including the folks who wrote the issues) really understand what it does.  I *honestly* think we should deprecate the whole thing.  I think if something has vanishingly few published examples (are there any in real journals?  I'd never let someone put this in a paper I was reviewing or editing), *and* it is subject to misunderstandings like this is, we should really reconsider its value.  
2) If we are going to keep it, `linthreshx/y` should be positional with no default so that people who are using it *need* to understand it.  If we did that, then the function would error on `linthresy`.  (It would also error on the default no-argument, but...)

**Comment by anntzer** [comment]
mostly agree on 1)
on 2) we can make linthresh{x,y} keyword-only but with no default

**Comment by ImportanceOfBeingErnest** [comment]
Mhh, let's deprecate `.plot()` then. It draws lines between data points, which results in visuals which are scientifically complete garbage.  
Or maybe just stop matplotlib development completely, given that each day thousands of students misuse it to draw nonsensical plots.

Back on topic: Just deprecate passing unused kwargs to this, or any function or method you like and everyone is happy. (Use a one-minor-release-cycle if you think it's urgent - no opposition on that from my side)

**Comment by jklymak** [comment]
>Mhh, let's deprecate .plot() then. It draws lines between data points, which results in visuals which are scientifically complete garbage. 

Given that there are hundreds of thousands of peer-reviewed articles that use `plot` (or its equivalent in other packages) I don't think the majority of the scientific community agrees that lines connecting data are "complete garbage".  I'm not saying symmetric log plots are complete garbage.  I'm just saying that I've *never* seen a symmetric log plot in a published paper.  If there is a domain-specific community that uses it, by all means they should implement it, or whatever scale they see fit.  Having the ability to define custom scales is a great functionality.  I just don't think every obscure scale should be part of *core* Matplotlib.

**Comment by anntzer** [comment]
@tacaswell You put in a similar API tightening in https://github.com/matplotlib/matplotlib/commit/63572455d605cf9f0c7ddbd1e9e785692492f64d without a deprecation period.  Would you support doing the same here (see discussion above)?  I will defer to your judgment.

**Comment by tacaswell** [comment]
We should all take it down a notch, we all want the same thing at the end of the day, just disagree a about what some details along the way are :)

TL; DR : we should do a deprecation cycle here,  I'll push some commits.

-----


@ImportanceOfBeingErnest is completely correct that this is an API change and should be documented and go through a deprecation cycle.

@anntzer is also completely correct that I set precedent of making this sort of API change in a patch release (!).

In my judgement @timhoffm and @anntzer are almost almost certainly correct that everyone we will catch with this change _intend_ to be passing a kwarg we do not ignore, however @ImportanceOfBeingErnest is also correct that there is a chance that there is some library layer which is abusing the current behavior (I could imaging having a dict of kwargs `{'linthreshy': 1.5, 'linthreshx': 1.5}` which `**kw` into both xscale and yscale) that we do not want to break without warning.

On balance, I think I made a mistake in 6357245 , and although we have not (to my knowledge) had any complaints about it, we should try to do better about not breaking down stream and users without warning.

Technically the same goes for changing the exception type, but I think we should let that one through as there is no reasonable way to do warnings on that (I guess MI of the two exceptions but then how do you tell when you are caught by the 'wrong one'?).

------

As a side note, these scales have been in from pre 1.0 days (via 2e8ce8170d8a3f972c505986bf3aeee47a8821e7 in 2007).  Log scales in the negative direction make perfect sense (you make it positive, take the log, put the negative backoutside), log has a linear region around 1  (`ln(1+x) ~ x`) so there is a reasonable place to smoothly go from log scale -> linear scale and using linear in the middle prevents a discontinuity as your data goes through zero if you have symmetric-around-0 multi-scale data.  If someone came today and offered a scale like this I would push them to make a stand alone package, but this scale (and `plot`) are not going anywhere anytime soon.
